src/main/java/duke/TaskList.java

```diff
@@ -0,0 +1,228 @@
+                        res.append("\n").append(i).append(".").append(t);
+                        i++;
+                    }
+                }
+                return res.toString();
+            } else {
+                return "No matching task has been found";
+            }
+        } else {
+            throw new InvalidCommandException();
+        }
+    }
+
+    /**
+     * Completes the task at the position in the list which the user specifies.
+     *
+     * @param str User input
+     * @return A string indicating completion of the task.
+     * @throws TaskCompletionException If the number is out of range of the list.
+     */
+    public String completeTask(String str) throws TaskCompletionException {
+        if (!str.startsWith("done ")) {
+            throw new TaskCompletionException(tasks.size());
+        }
+        String val = str.substring(5);
+        if (isInteger(val)) {
+            int i = Integer.parseInt(val);
+            if (i > 0 && i <= tasks.size()) {
+                if (tasks.get(i - 1).complete()) {
+                    return "Task is already completed.\n" + tasks.get(i - 1);
+                } else {
+                    return "Task completed: " + tasks.get(i - 1);
+                }
+            } else {
+                throw new TaskCompletionException(tasks.size());
+            }
+        } else {
+            throw new TaskCompletionException(tasks.size());
+        }
+    }
+
+    /**
+     * Deletes the task at the position in the list which the user specifies.
+     *
+     * @param str User input
+     * @return A string indicating deletion of the task.
+     * @throws TaskDeletionException If the number is out of range of the list.
+     */
+    public String deleteTask(String str) throws TaskDeletionException {
+        if (!str.startsWith("delete ")) {
+            throw new TaskDeletionException(tasks.size());
+        }
+        String val = str.substring(7);
+        if (isInteger(val)) {
+            int i = Integer.parseInt(val);
+            if (i > 0 && i <= tasks.size()) {
+                String res = "Task has been removed.\n" + tasks.get(i - 1);
+                delete(i);
+                res += "\nYou now have " + tasks.size() + " tasks in the list";
+                return res;
+            } else {
+                throw new TaskDeletionException(tasks.size());
+            }
+        } else {
+            throw new TaskDeletionException(tasks.size());
+        }
+    }
+
+    /**
+     * Deletes all items in the current list.
+     * @return A string indicating tasks have been cleared.
+     */
+    public String clear() {
+        tasks.clear();
+        return "Task List has been cleared.";
+    }
+
+    private String insert(Task task) {
+        tasks.add(task);
+        return "Task has been added:\n" + task + "\nYou now have " + tasks.size() + " tasks in the list";
+    }
+
+    private void delete(int i) {
+        Task t = tasks.get(i - 1);
```

Comment:
Perhaps this variable could be provided with a more descriptive name to improve readability.